### PR TITLE
Sav - Fix Profile Weekly Committed Hours Input

### PR DESCRIPTION
**Pull request by Lunaliar**
Fixes bug #12 Priority Medium  from USER PROFILE COMPONENT

# Description

On the profile page under the volunteering times tab, a user with edit permissions can change the weekly committed hours to unrealistic amounts. I've added a handle change on the input that should prevent users from adjusting those hours to anything below zero or anything above 168 (The maximum hours in one week). 

…

## Main changes
- Added handleChange function to check the status of values being passed to the input
- Give input min and max properties
- Alert user if an unrealistic value is attempted
…

## How to test:
1. Check into the current branch
2. Do `npm install` and `...` to run this PR locally
3. Log in as a user who can edit times on a profile (a user with canEdit: true)
4. Go to dashboard→ Profile→ Volunteering times→Weekly Committed Hours
5. Verify component WeeklyCommittedHours works properly and that users are not able to change their committed hours to an unrealistic number.

## Screenshots:
[PR MaxHours.webm](https://user-images.githubusercontent.com/98275089/224061105-1d618664-878e-4add-b821-38039c505df1.webm)

### CODE REVIEW

**Comment by nicollext** [comment]
Hi Sav! Good Job. 

Tested and it is working well. 


https://user-images.githubusercontent.com/101591708/224188063-65cad2e8-d40e-4401-b5b9-cfa4066e15f2.MP4

**Review by nicollext — DISMISSED** [review]

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx [hunk 1]
```diff
@@ -7,6 +7,9 @@ import { ENDPOINTS } from 'utils/URL';
 import axios from 'axios';
 import styles from './VolunteeringTimeTab.css';
 
+const MINIMUM_WEEK_HOURS = 0;
```
Comment (crystalsong2021) [inline src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:10]: I feel like this min_hour should start at 1 since it is the hour of volunteer time.

**Comment by crystalsong2021** [comment]
Nice job @Lunaliar . UI works as intended and I tried with letter and it doesn't allow entry until I type in the number.
on note: I noticed the warning sign on `canEdit` props. 

https://user-images.githubusercontent.com/18197013/224195778-20a5a733-bced-44d7-a406-b7397731f4fc.mov

Screenshot 2023-03-09 at 4 53 52 PM

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx (hunk 1 above)
Comment (Lunaliar) [reply src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:10]: Great point! I thought about this as well but I didn't know if there might be changes later on like a minimum of 5  or 10 being required so I just focused on not allowing negative numbers and setting up a constant that we can update later down the line when we're ready.

Comment (vitorsaa2k) [reply src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:10]: i think it should be 10 as the currently minimum is 10

**Comment by vitorsaa2k** [comment]
Working as intended, great job!

https://user-images.githubusercontent.com/17365161/224389292-00227a1d-cbfe-4702-885e-615c724eeba6.mp4

**Comment by mpavithr** [comment]
Hi, when testing this, when I set the weekly committed hours to over or under the allowed amount by manually typing it in, I get the pop-up.
pr704_1
When I try to go below the allowed minimum with the arrows, it doesn't allow me and when I go over the maximum with the arrows, it doesn't allow me either, but I get no pop-up.
pr704_2

**Review by Lucas-E — COMMENTED** [review]
Hey, just tested your PR and it works as intended, great job!
Captura de tela de 2023-03-11 02-21-27

**Review by kangjay88 — DISMISSED** [review]
Hi @Lunaliar, I have tested this PR and it works as intended. I mentioned this in PR#700, but it would be great if a future PR could make the spelling for 'committed' consistent so as not to confuse people making changes later on. Awesome work!
PR#704 Test
PR#704 Mobile

**Comment by Lunaliar** [comment]
> Hi @Lunaliar, I have tested this PR and it works as intended. I mentioned this in PR#700, but it would be great if a future PR could make the spelling for 'committed' consistent so as not to confuse people making changes later on. Awesome work! PR#704 Test PR#704 Mobile

Can't agree more, it tripped me up more than once while working on this, I believe I read in a previous PR that it was a back-end issue that we're going to work on soon!

**Comment by Lunaliar** [comment]
> Hi, when testing this, when I set the weekly committed hours to over or under the allowed amount by manually typing it in, I get the pop-up. pr704_1 When I try to go below the allowed minimum with the arrows, it doesn't allow me and when I go over the maximum with the arrows, it doesn't allow me either, but I get no pop-up. pr704_2

Good idea about the incremental changes, Ill take a look and see if I can get that working like you described thanks!

**Comment by Lunaliar** [comment]
> When the maximum hours is trigged it works as intended, great job!
> 
> But changing the variable `MINIMUM_WEEK_HOURS` to 10 and then trying to change the input value to less than 10(triggering the minimum hours) make the save warning appear. This happens because the warning is being handled with the setChange state. When the new handle for the warning(from PR #700 ) be merged this will not be an issue anymore.

Great Idea! I'll change it to 10 hrs.

**Comment by Lunaliar** [comment]
> Nice job @Lunaliar . UI works as intended and I tried with letter and it doesn't allow entry until I type in the number. on note: I noticed the warning sign on `canEdit` props.
> 
> Screenshot 2023-03-09 at 4 53 52 PM

Thanks! I also noticed the warning on canEdit as well but I figured it was outside of the scope of this PR. Hopefully, we can get that sorted out soon though!

**Comment by one-community** [comment]
Minimum hours should not be limited to 10, they should be able to be set as low as zero. Please confirm this is the case.

**Comment by Lunaliar** [comment]
> Minimum hours should not be limited to 10, they should be able to be set as low as zero. Please confirm this is the case.

No problem, I have it set to 0 already!

**Comment by NyahJ** [comment]
Hi @Lunaliar,

The test results from my side went as expected, other than the test steps mentioned above, I also tested it by the negative numbers input, there was a popup window shown like below, great job! Thank you!

截屏2023-03-11 下午9 59 40

**Comment by Jorge-Nerd** [comment]
Hi  @Lunaliar I work fine for me 

https://user-images.githubusercontent.com/77848827/225464627-72dc255f-e51d-428d-b9b9-c75e48783246.mp4

**Comment by AriaYu927** [comment]
Hi @Lunaliar , I reviewed your PR and it works perfect. Great job, thank you!


https://user-images.githubusercontent.com/94303659/225661130-2cf4aacb-d07f-4f6e-904d-5dae8740e3c8.mp4

**Comment by AriaYu927** [comment]
Hi @Lunaliar , I tested your PR with typing illegal numbers and changing the number to illegal by arrow. It works great under all conditions. Great job, thank you!.

PR#704-0
PR#704-168

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx [hunk 2]
```diff
@@ -85,17 +88,40 @@ const WeeklyCommitedHours = props => {
...
     return <p>{props.userProfile.weeklyComittedHours}</p>;
   }
+  const handleChange = e => {
+    // Maximum and minimum constants on lines 9 & 10
+    // Convert value from string into easy number variable
+    const value = parseInt(e.target.value);
+    if (value > MAXIMUM_WEEK_HOURS) {
+      // Check if Value is greater than total hours in one week
+      alert(`You can't commit more than ${MAXIMUM_WEEK_HOURS} hours per week.`);
+      if (value === 169) {
```
Comment (PrabhjotSembhi) [inline src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:98]: Why are you checking this? if value === 169 and setting maximum committed time to 168. 
Either you should do this for every value that is greater than 168 or don't do it.

What's the purpose of this ( please let me know If I am missing anything that I need to know )

Comment (Lunaliar) [reply src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:98]: Great point! I originally was doing something similar to what you suggested but I wanted the alerts to also pop up when users are using their arrow keys as well. I set this conditional to check if the value is 169 because otherwise, I want it to stay the same as what the user is attempting to enter. If I have it check for any value greater than 168 then if a user attempts to write 40 but accidentally types 400 their value will be set to 168. I only want the value to be reset to 168 if it's 169. This way users who are using arrow keys to go up or down will hit the alerts and have their values set to the minimums and maximums, while users who are typing in their values don't have to go through the trouble of retyping if they make a mistake. Let me know if you have any further questions or suggestions on how this can be done better!

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx
Comment (Vishvesh-Sheoran) [inline src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:1]: Loc 324 to 353, an improved version can be:

<React.Fragment key={`hours-by-category-${key}`}>
  <Row>
    <Col md="6">
      <Label>
        {key !== 'unassigned'
          ? `Total Tangible ${capitalize(key)} Hours`
          : `Total Unassigned Category Hours`}
      </Label>
    </Col>
    <Col md="6">
      {canEdit && (
        <Input
          type="number"
          id={`${key}Hours`}
          step=".01"
          value={parseFloat(userProfile.hoursByCategory[key])?.toFixed(2)}
          onChange={(e) => {
            const newHoursByCategory = {
              ...userProfile.hoursByCategory,
              [key]: Number(e.target.value),
            };
            setUserProfile((prevProfile) => ({
              ...prevProfile,
              hoursByCategory: newHoursByCategory,
            }));
            setChanged(true);
          }}
          placeholder={`Total Tangible ${capitalize(key)} Hours`}
        />
      )}
    </Col>
  </Row>
</React.Fragment>


In this improved version, the following changes are made:

1. The _key_ attribute in the _React.Fragment_ is wrapped with template literals.
2. The conditional rendering for the label is simplified using template literals.
3. The ternary operator is removed, and the strings are interpolated using template literals for readability.
4. The conditional rendering for the input field is simplified using the logical AND operator.
5. The _setUserProfile_ function is updated to use the functional form, ensuring that the latest state is used.
6. The _const_ variable _newHoursByCategory_ is introduced to make the code more readable.
7. These changes improve the readability and maintainability of the code.

**Review by abdel-lall — COMMENTED** [review]
works as intended. Nice job!

PR704

**Comment by EdwinLauM** [comment]
Hi @Lunaliar,

Tested going below 0, one above 168 and a bigger range of numbers and it works as intended! Good job! 

https://user-images.githubusercontent.com/42288987/236391088-512d65d9-2814-4684-abb6-4af55f400adc.mov

**Comment by one-community** [comment]
Thank you all, moving to final review.

**Comment by AuroraLovesWhales** [comment]
Hi @Lunaliar! I am using the latest development branch on both backend and frontend. When I tested this PR today, I ran into the following problem: after inputting an initial number in Weekly Committed Time, the following error occurred, and the page was frozen until I refreshed it: 

image

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/ea878700-2dad-444b-9312-4ea586db4ae3

**Review by GKD890 — COMMENTED** [review]
Hi Sav, thanks for resolving conflicts for easier reviewing.
But I found when I type in some value in the input, two errors happen. 

Frontend_PR#704_feedback_error


Other than that, the number refresh if using arrow key has become slower than last version. Also, after I double click to select value in the input, the whole input can't response unless I refresh the page. 
Check the record [here,](https://www.dropbox.com/s/mgsfeozbuo05mru/Frontend_PR%23704_no-response-after-click.mov?dl=0) you can see if you can reproduce the same one.

**Review by rauleffting — APPROVED** [review]
Great job! Approved!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/29555732/2ba290c8-ff6f-45d5-bb97-90f1495a6f01

File: src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx [hunk 3]
```diff
@@ -74,27 +85,47 @@ const WeeklySummaryOptions = props => {
...
+    } else {
+      props.setUserProfile({ ...props.userProfile, weeklyComittedHours: value });
+      props.setChanged(true);
+    }
+  };
+
   return (
     <Input
       type="number"
+      min={MINIMUM_WEEK_HOURS - 1}
```
Comment (cvtqx) [inline src/components/UserProfile/VolunteeringTimeTab/VolunteeringTimeTab.jsx:122]: Hi @Lunaliar - may I ask why you set min and max values this way and not just min={MINIMUM_WEEK_HOURS} max={MAXIMUM_WEEK_HOURS}?

**Comment by cvtqx** [comment]
When inputting value into the field manually, it works as described. However, when using up and down errors, the page freezes and the attached errors are displayed in console:
Screen Shot 2023-06-05 at 2 10 29 PM
